### ISSUE TITLE
Bug: [no-unnecessary-type-assertion] false positive when casting changes the inferred type from a generic

### ISSUE THREAD

**Issue by DetachHead** [typescript-eslint/typescript-eslint#5359]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&code=CYUwxgNghgTiAEYD2A7AzgF3gMyUgXPADwAqANPAKoB8AFAG5QQCuIhlAlPALzXwkBYAFDCA9KIDmIFCBgBLMGniwEKJAHd4clNllxgypURTMAtgCNZFTPJQS6jFm3g3tErr3gmLs4cnRY5rA8OHi0AEThXFBK3pYw8OJ6SDCEagC0zCgyYCBoaLAAnukYhQAOIOkxaLIYcqhAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
declare const foo: <T, U>(value: U) => T

//generics are now inferred as <number, string>(value: string) => number
const bar = foo("") as number //error: no-unnecessary-type-assertion
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  }
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

no error

### Actual Result

error: no-unnecessary-type-assertion

### Additional Info

sometimes it's useful to use a cast instead of specifying the generic, in case you only want to have to specify one type while still having the other types inferred.

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `5.30.7` |
| `@typescript-eslint/parser`            | `5.30.7` |
| `TypeScript`                           | `4.7.2` |
| `ESLint`                               | `8.15.0` |

**Comment by Josh-Cena** [comment]
Try `document.querySelectorAll<HTMLImageElement>("")`? Assertions are inference sites so they impact the generic function.

**Comment by DetachHead** [comment]
that worked, thanks! interesting i had no idea casting effected generic inference.

in that case i've updated the issue with a better example

**Comment by bradzacher** [comment]
Yup, Josh is right.

Ehen a generic us only used in tge return type then TS will propagate the asserting into the generic.

Using a generic in a return type is generally a code smell and breaks the golden rule of generics (a generic in a function should be used at least twice) https://effectivetypescript.com/2020/08/12/generics-golden-rule/

---

Check out thus example which shows hiw TS "backfills" the generic based on the assertion

```ts
declare function foo<T>(): T;

foo<string>();
foo() as string;
```
[playground](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtVJxwB4AVAPgAoBKALnhIG4AoJ+A4gZwxi1QHNKVZgHph8AHoB+Vu2rwoHeFx78RYqUyA)
